### ISSUE TITLE
[Bug] how the runner resgister the project module

### ISSUE THREAD

**Issue by peiliu0408** [open-mmlab/mmengine#1100]
### Prerequisite

- [X] I have searched [Issues](https://github.com/open-mmlab/mmengine/issues) and [Discussions](https://github.com/open-mmlab/mmengine/discussions) but cannot get the expected help.
- [X] The bug has not been fixed in the latest version(https://github.com/open-mmlab/mmengine).

### Environment

two modules, i.e., mmaction2 and mmpretrain (both based on mmengine), were put under my own project for a more convenient experiment env.
If only model from mmacion2 is conduct in experment, it works perfect. Same as model from mmpretrain.

But, two model, one from mmaction2 (like convenient, support image, video as inputs), another from mmpretrain (i.e., the ImageNet dataset module). An unexpected errors occurs:

ImageNet is not resgisterd.

Obviously, the whole mmpretrained is not resgisterd. 

As introduced in doc, mmengine manage modules in Register  class using "scope". 

But how the Runner automatically register more than on module (mmaction, mmpretrain) automatically?
1. how the Runner automaticaly register the module, like mmaction?
2. whether the Runner support register more than one modules at the same time?

### Reproduces the problem - code sample

the project is organized like this:

configs -- copy from mmaction2 project
mmaction -- copy form mmaction2 project
mmpretrain -- copy from mmpretrain project
train.py
test.py
other unrelated files

### Reproduces the problem - command or script

a similar config: https://github.com/open-mmlab/mmaction2/blob/dev-1.x/configs/recognition/omnisource/slowonly_r50_8xb16-8x8x1-256e_imagenet-kinetics400-rgb.py

ImageNet dataset is registerd in mmpretrain proejct but not in mmaction

### Reproduces the problem - error message

the ImageNet dataset is not found.

**Comment by HAOCHENYE** [comment]
I guess this problem is caused by the concept: "scope". Each repo defines their [`default_scope`](https://github.com/open-mmlab/mmaction2/blob/516700439705fc56df0a86e33916e7a26c55837a/configs/_base_/default_runtime.py#L1) in this `default_runtime.py`. and this variable will be parsed in Runner. 

The `scope` will influence the behavior of how `Registry` search the module. For example, if `scope` is `mmaction`, the `Registry` will search the module in the `Registry` object defined [here](https://github.com/open-mmlab/mmaction2/blob/main/mmaction/registry.py) . I think the error:

`ImageNet dataset is registerd in mmpretrain proejct but not in mmaction`

is caused by the scope is not set correctly in your config. If you want to search the module from the registry of `MMPretrain`, there are two ways to  do it:

```
dataset = dict(
    type='mmpretrain.ImageNet'
)
```
or 
```
dataset  = dict(
    _scope_='mmpretrain',
    type='ImageNet'
)
```

You can find more information about scope [here](https://mmengine.readthedocs.io/en/latest/advanced_tutorials/registry.html#getting-started)

**Comment by peiliu0408** [comment]
Yeah, as mentioned in office doc, _cross scope calling is a new featue from mmengine_.

I think I did not make my problem clear.
I start the code with the config from mmaction (i.e., the model, dataset, pipeline, all related modules are defined in mmaction).

_p.s. there is a scope related setting in config file: 
`default_scope = 'mmaction'`_
https://github.com/open-mmlab/mmaction2/blob/5ed09fd3bf290fadab5458645e029ef66cd876c4/configs/_base_/default_runtime.py#L1

In my work, I want to **joint** traing a model with **image and video** data, **therefore, ImageNet dataset, defined in mmpretrain, is adopted in a cross scope manner**.

situation 1:
**I installed mmpretrain, my current problem could be solved**. **_BUT_**,
I start the an another image task, such as an image classification task,
`error: the **** has been registed, `
Therefore, I think the Runner would register a folder module (mmaction or mmpretrain) automatically. When going deeper, i think this registration work is done here in runner.py: 
`self.log_processor = self.build_log_processor(log_processor)`

https://github.com/open-mmlab/mmengine/blob/9868131c98c1f3bf9eb41930e671b43c71f55cb9/mmengine/runner/runner.py#L374

this conclusion is got by using tool `count_registered_modules()` provided by mmengine. **That weird,  the module registration is done when calling build_log_process func.**

OK, anyway, the mmaciton module is registed by runner.

situation 2: do not install mmpretrain.
I debug into the line in runner.py, `self.log_processor = self.build_log_processor(log_processor)`,
before executing: found only mmengine in registered_modules, 
aftrer executing: mmengine, mmaction module is registered.

The problem, I can I register the mmaction and mmpretrain at the same time in a direct way ?

**Comment by HAOCHENYE** [comment]
Actually, the register process is triggered in **Registry.get** 

https://github.com/open-mmlab/mmengine/blob/cdec4cbd4acbabca6cc3bc7bb0866b9a67d8f401/mmengine/registry/registry.py#L373

and here

https://github.com/open-mmlab/mmengine/blob/cdec4cbd4acbabca6cc3bc7bb0866b9a67d8f401/mmengine/registry/registry.py#L377

In most of the cases, `Registry.get` is called by [`build_from_cfg`](https://github.com/open-mmlab/mmengine/blob/cdec4cbd4acbabca6cc3bc7bb0866b9a67d8f401/mmengine/registry/build_functions.py#L98).
`build_from_cfg` is called anywhere, the code snipped you mentioned `self.build_log_processorr` will also step into the
`build_from_cfg`. But what `build_from_cfg` does?

1. **Switch to the target registry**: `build_from_cfg` is commonly called by a `Regitry` object, such as `DATASETS` defined in MMEngine. However, MMEngine does not implement any specific dataset, and the dataset is actually implemented and registered into the registry of downstream repos. Therefore, MMEngine needs to [switch](https://github.com/open-mmlab/mmengine/blob/cdec4cbd4acbabca6cc3bc7bb0866b9a67d8f401/mmengine/registry/build_functions.py#L95) to the `Registry` from `MMEngine` to `MMPretrain` according to the `scope`(scope is initialized [here](https://github.com/open-mmlab/mmengine/blob/cdec4cbd4acbabca6cc3bc7bb0866b9a67d8f401/mmengine/runner/runner.py#L370)) which is defined in `default_runtime.py` for searching the target class.

2. **Trigger the registering**: For example, now we've got the registry `DATASETS` defined in `MMPretrain`, we need to trigger the registering. As described before, the `Registry.get` is responsible for doing this.
3. **Search the class**: Commonly, `Registry` will search the class from the current registry (`mmpretrain.registry.DATASETS`) and the root registry (`mmengine.registry.root.DATASETS`). **However**, if you define your `type` like `mmpose.xxxDataset`, it will further switch to the registry of `mmpose` and search the target class.
4. **Built the target object**

Back to the problem you meet, `build_log_process` will actually trigger the registering since it will call `build_from_cfg` and switch to the registry of mmpretrain and register all modules.

Consider you want to use mmpretrain and mmpose jointly, you need to consider the following things:

1. What should the default scope be in your project? The default scope will determine to search the target class from which registry by default, **mmpose** or **mmpretrain**.
2. If you've chosen your default scope, for example,  mmpretrain, and then you need to define the `type` from mmpose like `mmpose.xxx` rather than simply xxx

**Comment by peiliu0408** [comment]
thanks for your reply, and it more clear foo understanding the scope system. The purpose of designing scope actually is to bridge isolated repos, and bring a more convienient env for developers, let modules like Lego block ignoring definition position. Unfortunatly, many details are hided, and office doc does not give an illustration.

as mention in above, **3. Search the class**
can you give the corresponding search **PATH**, it's a directory list or something else?

**Comment by HAOCHENYE** [comment]
A simple rule is that you need to make sure `import mmpose` or `import mmpretrain` will not raise an error. You need to install `mmpose` or `mmpretrain` by `pip instal xxx` or `pip install -e .`(install from source). Alternatively, you can also make them importable by setting PYTHONPATH like `export PYTHONPATH=${path/to/pose}:$PYTHONPATH`

**Comment by peiliu0408** [comment]
thanks a lot. that finaly works in a elegant manner.
